File: frontend/lib/qcEffort.ts
/** QC effort from real Test Cases. Keep in sync with backend `qc_effort`. */

export const QC_HOURS_PER_DAY = 6;
export const BLOCKER_MINUTES = 20;
export const CRITICAL_MINUTES = 12;
export const COMPLEXITY_FACTOR: Record<string, number> = {
  BAJA: 1.0,
  LOW: 1.0,
  MEDIA: 1.3,
  MEDIUM: 1.3,
  ALTA: 1.6,
  HIGH: 1.6,
};

/** Retired count formula, comparison only: (TC / 46) × 3 days. */
export const QC_CASES_PER_DAY_LEGACY = 46;
export const QC_RELEASE_EFFORT_FACTOR_LEGACY = 3.0;

export const QC_OPERATIVA_CASES_PER_DAY = 6;

export const QC_ESTIMATION_TOOLTIP =
  "Esfuerzo QC = minutos base (BLOCKER 20 / CRITICAL 12) × factor (BAJA 1.0 / MEDIA 1.3 / ALTA 1.6). La complejidad sale de pasos/condición/confianza, no de la prioridad. Días-persona = horas / 6. Duración = días-persona / recursos. La ventana de ejecución es calendario y no entra en esta fórmula.";

export const QC_OPERATIVA_ESTIMATION_TOOLTIP =
  "Operativa: 1 tester por dispositivo. Horas = N × (6 h/día ÷ 6 TC/día) = N × 1 h. Días QC = max(casos del dispositivo más cargado) ÷ 6 TC/día (trabajo en paralelo). Filtra un dispositivo para ver tu slice.";

export function priorityBaseMinutes(priority?: string | null): number {
  return String(priority || "").toUpperCase() === "BLOCKER" ? BLOCKER_MINUTES : CRITICAL_MINUTES;
}

export function complexityFactor(complexity?: string | null): number {
  const key = String(complexity || "MEDIA").trim().toUpperCase();
  return COMPLEXITY_FACTOR[key] ?? COMPLEXITY_FACTOR.MEDIA;
}

export function estimateCaseMinutes(priority?: string | null, complexity?: string | null): number {
  return Math.round(priorityBaseMinutes(priority) * complexityFactor(complexity) * 10) / 10;
}

export function estimateReleaseEffortFromCases(
  cases: Array<{ priority?: string | null; complexity?: string | null }>,
): { hours: number; days: number; minutes: number } {
  if (!cases.length) {
    return { hours: 0, days: 0, minutes: 0 };
  }
  const minutes = cases.reduce((sum, row) => sum + estimateCaseMinutes(row.priority, row.complexity), 0);
  const realHours = minutes / 60;
  const days = realHours / QC_HOURS_PER_DAY;
  return {
    minutes,
    hours: Math.round(realHours * 10) / 10,
    days: Math.round(days * 10) / 10,
  };
}

export function durationDays(personDays: number, resources: number): number {
  const testers = Math.max(1, Math.floor(resources) || 1);
  return Math.round((personDays / testers) * 10) / 10;
}

export function estimateReleaseEffortLegacyCount(testCaseCount: number): { hours: number; days: number } {
  const count = Math.max(0, Math.floor(testCaseCount));
  if (count === 0) {
    return { hours: 0, days: 0 };
  }
  const days = (count / QC_CASES_PER_DAY_LEGACY) * QC_RELEASE_EFFORT_FACTOR_LEGACY;
  const hours = days * QC_HOURS_PER_DAY;
  return {
    hours: Math.round(hours * 10) / 10,
    days: Math.round(days * 10) / 10,
  };
}

export function estimateOperativaEffort(cases: Array<{ device?: string | null }>): { hours: number; days: number } {
  const count = cases.length;
  if (count === 0) {
    return { hours: 0, days: 0 };
  }
  const byDevice = new Map<string, number>();
  for (const row of cases) {
    const key = (row.device || "").trim() || "Sin dispositivo";
    byDevice.set(key, (byDevice.get(key) || 0) + 1);
  }
  let busiest = 0;
  for (const n of byDevice.values()) {
    busiest = Math.max(busiest, n);
  }
  const hoursPerCase = QC_HOURS_PER_DAY / QC_OPERATIVA_CASES_PER_DAY;
  const hours = Math.round(count * hoursPerCase * 10) / 10;
  const days = Math.round((busiest / QC_OPERATIVA_CASES_PER_DAY) * 10) / 10;
  return { hours, days };
}

export function stripDeviceFromCaseName(name: string, device?: string | null): string {
  const label = (device || "").trim();
  if (!name || !label) {
    return name;
  }
  const escaped = label.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  return name.replace(new RegExp(`(?:\\s*[·•|]\\s*|\\s+-\\s+)${escaped}\\s*$`, "i"), "").trim() || name;
}

File: backend/app/services/qc_effort.py
"""QC effort from persisted Test Cases: priority minutes × complexity factor.

minutes = base(BLOCKER 20 / CRITICAL 12) × factor(BAJA 1.0 / MEDIA 1.3 / ALTA 1.6)

Calibrated to Coship 52: QC executed in ~2–2.5 days vs 4.7 with the previous 25×1.5 floor.
Complexity is classified from steps, condition and confidence only.
"""

from __future__ import annotations

from typing import Any, Literal

from app.config import settings
from app.schemas.case_generation import GeneratedCaseCandidate

Complexity = Literal["BAJA", "MEDIA", "ALTA"]

QC_HOURS_PER_DAY = 6.0
BLOCKER_MINUTES = 20
CRITICAL_MINUTES = 12
COMPLEXITY_FACTOR = {
    "BAJA": 1.0,
    "LOW": 1.0,
    "MEDIA": 1.3,
    "MEDIUM": 1.3,
    "ALTA": 1.6,
    "HIGH": 1.6,
}

# Retired count formula, kept only so reports can compare old vs new.
QC_CASES_PER_DAY_LEGACY = 46.0
QC_RELEASE_EFFORT_FACTOR_LEGACY = 3.0


def hours_per_day() -> float:
    return float(settings.qc_hours_per_day)


def classify_complexity(candidate: GeneratedCaseCandidate) -> Complexity:
    """Complejidad por forma del caso, no por prioridad.

    BAJA: validación básica de 1 paso, o 1 paso con confianza alta.
    MEDIA: condición especial, 2–3 pasos, o confianza media.
    ALTA: 4+ pasos o confianza baja.
    BLOCKER/CRITICAL no entran aquí; solo fijan la base de minutos (20 vs 12).
    """
    steps = len(candidate.steps)
    if candidate.basic_validation and steps <= 1 and not candidate.requires_condition:
        return "BAJA"
    if steps >= 4 or candidate.confidence == "low":
        return "ALTA"
    if candidate.requires_condition or steps >= 2 or candidate.confidence == "medium":
        return "MEDIA"
    return "BAJA"


def _priority_value(priority: Any) -> str:
    if priority is None:
        return "CRITICAL"
    return str(priority.value if hasattr(priority, "value") else priority).upper()


def priority_base_minutes(priority: Any) -> int:
    return BLOCKER_MINUTES if _priority_value(priority) == "BLOCKER" else CRITICAL_MINUTES


def complexity_factor(complexity: str | None) -> float:
    key = (complexity or "MEDIA").strip().upper()
    return COMPLEXITY_FACTOR.get(key, COMPLEXITY_FACTOR["MEDIA"])


def estimate_case_minutes(priority: Any, complexity: str | None) -> float:
    return round(priority_base_minutes(priority) * complexity_factor(complexity), 1)


def estimate_case_hours(priority: Any, complexity: str | None) -> float:
    return round(estimate_case_minutes(priority, complexity) / 60.0, 2)


def estimate_release_from_cases(cases: list[Any]) -> tuple[float, float]:
    """Return (hours, person-days) from real TCs. days = hours / 6."""
    if not cases:
        return 0.0, 0.0
    minutes = 0.0
    for row in cases:
        priority = getattr(row, "priority", None)
        if isinstance(row, dict):
            priority = row.get("priority")
            complexity = row.get("complexity")
        else:
            complexity = getattr(row, "complexity", None)
        minutes += estimate_case_minutes(priority, complexity)
    hours = minutes / 60.0
    days = hours / hours_per_day()
    return round(hours, 1), round(days, 1)


def duration_days(person_days: float, resources: int) -> float:
    testers = max(int(resources), 1)
    return round(person_days / testers, 1)


def estimate_release_legacy_count(test_case_count: int) -> tuple[float, float]:
    """Retired: (TC / 46) × 3 days; hours = days × 6. Comparison only."""
    count = max(int(test_case_count), 0)
    if count == 0:
        return 0.0, 0.0
    days = (count / QC_CASES_PER_DAY_LEGACY) * QC_RELEASE_EFFORT_FACTOR_LEGACY
    hours = days * QC_HOURS_PER_DAY
    return round(hours, 1), round(days, 1)
